text/0000-extensible-enums-and-structs.md

```diff
@@ -0,0 +1,125 @@
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Introduce a notation for declaring an enum to be *extensible*, meaning
+that more variants or fields may be added in the future. This notation
+does not affect the use of the enum within the current crate, but it
+prevents downstream crates from employing features (such as exhaustive
+matches or struct literals) that would fail to compile if new
+variants/fields were added.
+
+# Motivation
+
+Imagine that one is writing a public-facing API that includes an
+`Error` enum (this exact scenario arises in libstd with the I/O error
+enum):
+
+```rust
+pub enum Error {
+    NoSuchFile,
+    NoPermissions,
+}
+```
+
+Providing an enum listing out the possible sources of error is a very
+convenient API for your users, because they can be use a `match`
+statement to easily identify and handle particular kinds of
+errors. Unfortunately, it's rather limiting on the future evolution of
+your library: if you wish to add a third kind of error in the future,
+you will be potentially breaking downstream code. After all, it is
+possible that some user was doing an exhaustive match against all
+possible sources of error:
+
+```rust
+match error {
+    Error::NoSuchFile => ...,
+    Error::NoPermissions => ...,
+}
+```
+
+To resolve this dilemna, this RFC proposes permitting enums to be declared
+as *extensible*:
+
+```rust
+pub enum Error {
+    NoSuchFile,
+    NoPermissions,
+    ..
+}
+```
+
+Extensible enums from other crates can never be exhaustively
+matched. This means that a downstream user attempting to match against
+an error from your library would have to include a wildcard option:
+
+```rust
+// from an external crate
+match error {
+    Error::NoSuchFile => ...,
+    Error::NoPermissions => ...,
+    _ => ...,
+}
+```
+
+Due to the presence of this wildcard, you can safely add variants
+without fear of breaking source compatibility with downstream clients.
+
+Note that extensibility is ignored within the current
+crate. Therefore, your library may internally write a match that
+exhaustively covers all sources of error. You don't need to worry
+about source compatibility with yourself, after all.
+
+# Detailed design
+
+Enum grammar is extended to permit a list of variants to be terminated
+with `..`. An enum declared with `..` is considered *extensible*.
+
+Extensible enums that are local to the current crate are treated the
+same as any other enum. Exhaustiveness checking is modified so that
+extensible enums that are not local to the current crate are never
+considered completely covered. (As if there was one additional variant
+beyond those declared, essentially.)
+
+Extensibility does not currently affect the enum in any other way. The
+representation in particular remains unchanged. This RFC does not
+attempt to address binary-level compatibility, only source
+compatibility.
+
+# Drawbacks
+
+More language syntax.
+
+# Alternatives
+
```

Comment:
On Tue, Feb 03, 2015 at 02:39:32AM -0800, Cristi Cobzarenco wrote:
> I also buy the consistency argument: in particular since moving modules between crates should really not break pattern matching expressions. Maybe place that relaxation on modules rather than crates?

I'm not sure why you would think this. There are already things that
are affected by moving between crates: impls come to mind.